tstate/tstate.go

```diff
@@ -93,61 +91,58 @@ func (ts *TState) getValue(_ context.Context, key string) ([]byte, bool, bool) {
 // FetchAndSetScope then sets the scope of ts to [keys]. If a key exists in
 // ts.fetchCache set the key's value to the value from cache.
 func (ts *TState) FetchAndSetScope(ctx context.Context, keys [][]byte, db Database) error {
-	ts.scopeStorage = map[string][]byte{}
+	ts.scopeStorage = make(map[string][]byte, len(keys))
+	ts.scope = make(map[string]struct{}, len(keys))
 	for _, key := range keys {
-		k := string(key)
-		if val, ok := ts.fetchCache[k]; ok {
+		ts.scope[string(key)] = struct{}{}
+		if val, ok := ts.fetchCache[string(key)]; ok {
 			if val.Exists {
-				ts.scopeStorage[k] = val.Value
+				ts.scopeStorage[string(key)] = val.Value
 			}
 			continue
 		}
 		v, err := db.GetValue(ctx, key)
 		if errors.Is(err, database.ErrNotFound) {
-			ts.fetchCache[k] = &cacheItem{Exists: false}
+			ts.fetchCache[string(key)] = &cacheItem{Exists: false}
 			continue
 		}
 		if err != nil {
 			return err
 		}
-		ts.fetchCache[k] = &cacheItem{Value: v, Exists: true}
-		ts.scopeStorage[k] = v
+		ts.fetchCache[string(key)] = &cacheItem{Value: v, Exists: true}
+		ts.scopeStorage[string(key)] = v
 	}
-	ts.scope = keys
 	return nil
 }
 
 // SetReadScope sets the readscope of ts to [keys].
 func (ts *TState) SetScope(_ context.Context, keys [][]byte, storage map[string][]byte) {
-	ts.scope = keys
+	ts.scope = make(map[string]struct{}, len(keys))
+	for _, key := range keys {
+		ts.scope[string(key)] = struct{}{}
+	}
 	ts.scopeStorage = storage
 }
 
 // checkScope returns whether [k] is in ts.readScope.
 func (ts *TState) checkScope(_ context.Context, k []byte) bool {
-	for _, s := range ts.scope {
-		// TODO: benchmark and see if creating map is worth overhead
```

Comment:
this very much depends on the size of the set. since lookup is O(n) if its reasonable to have even 8+ keys its probably worth it.  I aded the change but we can revert the commit if we feel the set size would remain similar to tokenvm (4)

Reply:
Although the lookup is much faster for map it does increase the alloc in `FetchAndSetScope` by a fair amount. As this set grows into possibly hundreds the time gained on lookup will be probably overshadowed by memory bloat/cpu time.

### before
```
goos: linux
goarch: amd64
pkg: github.com/ava-labs/hypersdk/tstate
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkFetchAndSetScope/fetch_and_set_scope_4_keys_with_length_65-12         	  894080	      1408 ns/op	     800 B/op	       7 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_8_keys_with_length_65-12         	  687463	      2001 ns/op	    1473 B/op	      11 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_16_keys_with_length_65-12        	  372249	      3031 ns/op	    2833 B/op	      19 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_32_keys_with_length_65-12        	  243076	      5262 ns/op	    5424 B/op	      35 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_64_keys_with_length_65-12        	  110935	     14345 ns/op	   11416 B/op	      68 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_128_keys_with_length_65-12       	   42150	     26776 ns/op	   22680 B/op	     132 allocs/op
```

### after
```
goos: linux
goarch: amd64
pkg: github.com/ava-labs/hypersdk/tstate
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkFetchAndSetScope/fetch_and_set_scope_4_keys_with_length_65-12         	  716866	      1802 ns/op	    1392 B/op	      14 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_8_keys_with_length_65-12         	  404772	      2923 ns/op	    2529 B/op	      22 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_16_keys_with_length_65-12        	  282474	      4318 ns/op	    4824 B/op	      38 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_32_keys_with_length_65-12        	  157590	      7921 ns/op	    9284 B/op	      70 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_64_keys_with_length_65-12        	   83341	     15155 ns/op	   19376 B/op	     136 allocs/op
BenchmarkFetchAndSetScope/fetch_and_set_scope_128_keys_with_length_65-12       	   42186	     40291 ns/op	   38448 B/op	     264 allocs/op
```

Reply:
reverting for now.

Reply:
You can see the cost of lookup as the set increases.
```
BenchmarkCheckScope/set_scope_4_keys_with_length_65-12         	55952859	        21.66 ns/op	       0 B/op	       0 allocs/op
BenchmarkCheckScope/set_scope_8_keys_with_length_65-12         	30695019	        39.45 ns/op	       0 B/op	       0 allocs/op
BenchmarkCheckScope/set_scope_16_keys_with_length_65-12        	14245750	        84.93 ns/op	       0 B/op	       0 allocs/op
BenchmarkCheckScope/set_scope_32_keys_with_length_65-12        	 7016282	       160.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkCheckScope/set_scope_64_keys_with_length_65-12        	 3868280	       286.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkCheckScope/set_scope_128_keys_with_length_65-12       	 2166471	       547.1 ns/op	       0 B/op	       0 allocs/op
```